supervisor.go

```diff
@@ -0,0 +1,339 @@
+package capataz
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+type routineInfo struct {
+	spec RoutineSpec
+	rs   *routineSupervisor
+}
+
+type SupervisorSpec struct {
+	MaxRestarts      uint
+	MaxRestartWindow time.Duration
+	RoutineSpecs     []RoutineSpec
+	RestartStrategy  RestartStrategy
+	Restart          Restart
+	StartOrder       Order
+	StopOrder        Order
+}
+
+type SupervisorSpecOption func(ss *SupervisorSpec)
+
+func WithMaxRestarts(n uint) SupervisorSpecOption {
+	return func(ss *SupervisorSpec) {
+		ss.MaxRestarts = n
+	}
+}
+
+func WithMaxRestartWindow(d time.Duration) SupervisorSpecOption {
+	return func(ss *SupervisorSpec) {
+		ss.MaxRestartWindow = d
+	}
+}
+
+func WithRoutineSpecs(specs []RoutineSpec) SupervisorSpecOption {
+	return func(ss *SupervisorSpec) {
+		ss.RoutineSpecs = specs
+	}
+}
+
+func WithRestartStrategy(strategy RestartStrategy) SupervisorSpecOption {
+	return func(ss *SupervisorSpec) {
+		ss.RestartStrategy = strategy
+	}
+}
+
+func WithRestart(restart Restart) SupervisorSpecOption {
+	return func(ss *SupervisorSpec) {
+		ss.Restart = restart
+	}
+}
+
+func WithStartOrder(order Order) SupervisorSpecOption {
+	return func(ss *SupervisorSpec) {
+		ss.StartOrder = order
+	}
+}
+
+func WithStopOrder(order Order) SupervisorSpecOption {
+	return func(ss *SupervisorSpec) {
+		ss.StopOrder = order
+	}
+}
+
+type CountResult struct {
+	Specs       int
+	Active      int
+	Supervisors int
+	Workers     int
+}
+
+type Supervisor interface {
+	// CountRoutines() CountResult
+	// DeleteChild(id string) error
+	// GetChildSpec(id string) (ChildSpec, error)
+	// RestartChild(id string) error
+	// StartChild(childSpec ChildSpec) error
+	// TerminateChild(id string) error
+}
+
+type supervisor struct {
+	sync.RWMutex
+	spec              SupervisorSpec
+	routineSpecs      []RoutineSpec
+	routines          map[string]*routineInfo
+	cancelContext     context.CancelFunc
+	restartTracker    restartTracker
+	wg                sync.WaitGroup
+	monitoringStarted bool
+}
+
+func supervisorFactory(spec SupervisorSpec) RoutineFactory {
+	return func() (Routine, error) {
+		return newSupervisor(spec)
+	}
+}
+
+func newSupervisor(spec SupervisorSpec) (*supervisor, error) {
+	s := &supervisor{
+		spec:         spec,
+		routineSpecs: cloneRoutineSpecs(spec.RoutineSpecs),
+		routines:     make(map[string]*routineInfo),
+		restartTracker: restartTracker{
+			max:    spec.MaxRestarts,
+			window: spec.MaxRestartWindow,
+		},
+	}
+
+	s.spec.StartOrder(s.routineSpecs, func(spec RoutineSpec) error {
+		return s.startRoutine(spec)
+	})
+
+	return s, nil
+}
+
+func (s *supervisor) Go(ctx context.Context) error {
+	if s.monitoringStarted {
+		return nil
+	}
+
+	return s.startMonitoring(ctx)
+}
+
+func (s *supervisor) startMonitoring(ctx context.Context) error {
+	s.monitoringStarted = true
+	s.wg.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	s.cancelContext = cancel
+
+	<-ctx.Done()
```

Comment:
If we get a cancel from a child, how do we get the reason this supervisor failed? `ctx.Err` would mask the real reason (error window/intensity got reached by worker `<worker-id>`)